### PR TITLE
Reorganize `patch` and `resource` blocks

### PR DESCRIPTION
**Pull request by Rylan12**
Sorry, another large PR. However, this one should be simpler and hopefully the last one.

This PR re-organizes the `patch` and `resource` blocks so that the `on_{system}` blocks are nested inside the `patch` and `resource` blocks. This allows these blocks to be placed in the correct places, instead of needing to be where the `on_{system}` blocks are. This will be enforced by rubocops once this is merged (see https://github.com/Homebrew/brew/pull/13539)

### CODE REVIEW

File: Formula/lua.rb [hunk 1]
```diff
@@ -42,6 +33,15 @@ class Lua < Formula
...
   end
 
+  # Be sure to build a dylib, or else runtime modules will pull in another static copy of liblua = crashy
+  # See: https://github.com/Homebrew/legacy-homebrew/pull/5043
+  patch do
+    on_macos do
+      url "https://raw.githubusercontent.com/Homebrew/formula-patches/11c8360432f471f74a9b2d76e012e3b36f30b871/lua/lua-dylib.patch"
+      sha256 "a39e2ae1066f680e5c8bf1749fe09b0e33a0215c31972b133a73d43b00bf29dc"
+    end
+  end
```
Comment (carlocab) [inline Formula/lua.rb:43]: Can we move the Linux patch into this block?

Comment (Rylan12) [reply Formula/lua.rb:43]: Oh yeah, good idea. I already found another thing to fix, so I'll do another run-through to find these things.

Comment (Rylan12) [reply Formula/lua.rb:43]: @carlocab, do you think we should just require all `on_{system}` blocks to be nested inside `resource` blocks, even when there are multiple?

Currently, cases like `kpcli` allow this:

```ruby
on_linux do
  resource "foo"
    ...
  end
  resource "bar"
    ...
  end
end
resource "baz"
  ...
end
```

Instead, we could require:
```ruby
resource "foo"
  on_linux do
    ...
  end
resource "bar"
  on_linux do
    ...
  end
end
resource "baz"
  ...
end
```

Comment (Rylan12) [reply Formula/lua.rb:43]: Also, with the `lua` example, we could combine them but I'm not sure we should since the patches are for different issues

**Comment by Rylan12** [comment]
Okay, I've gone through the formulae and merged `resource` blocks when I thought it was appropriate. There were a few `patch` blocks that I didn't merge because they were fixing separate issues, and it feels more appropriate to have each `patch` block be for its own issue.

One exception is the formula `llnode`. The `lldb` resource is needed on both macOS and Linux, but there are some `if DevelopmentTools.clang_build_version` conditions that make it more complex, and there are no `sha256` blocks since the resources use a `revision` instead. Something about this specific way means that `brew style` doesn't like it if we combine the `on_macos` and `on_linux` blocks into one formula, so I left it as-is for now. I'm not sure its worth fixing in this case since it's an obscure case with one instance, and I'm not immediately sure how to fix it.

File: Formula/lua.rb (hunk 1 above)
Comment (carlocab) [reply Formula/lua.rb:45]: Why do you say that? The macOS and Linux patches are both for building a shared `liblua`.

File: Formula/lanraragi.rb [hunk 2]
```diff
@@ -33,8 +33,8 @@ class Lanraragi < Formula
 
   uses_from_macos "libarchive"
 
-  on_macos do
-    resource "libarchive-headers" do
+  resource "libarchive-headers" do
```
Comment (carlocab) [inline Formula/lanraragi.rb:36]: Could probably be done in a separate PR, but it seems much more maintainable to use Homebrew `libarchive` than to workaround any missing headers by downloading them.

**Review by carlocab — APPROVED** [review]

**Pull request closed without merging**